File: objectModel/TypeScript/Persistence/CdmFolder/ConstantEntityPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmFolder } from '..';
import {
    CdmConstantEntityDefinition,
    CdmCorpusContext,
    cdmObjectType,
    copyOptions,
    resolveOptions
} from '../../internal';
import {
    ConstantEntity,
    EntityReferenceDefinition
} from './types';

export class ConstantEntityPersistence {
    public static fromData(ctx: CdmCorpusContext, object: ConstantEntity): CdmConstantEntityDefinition {
        const constantEntity: CdmConstantEntityDefinition =
            ctx.corpus.MakeObject<CdmConstantEntityDefinition>(cdmObjectType.constantEntityDef, object.constantEntityName);
        if (object.explanation) {
            constantEntity.explanation = object.explanation;
        }
        if (object.constantValues) {
            constantEntity.constantValues = object.constantValues;
        }
        constantEntity.entityShape = CdmFolder.EntityReferencePersistence.fromData(ctx, object.entityShape);

        return constantEntity;
    }

    public static toData(instance: CdmConstantEntityDefinition, resOpt: resolveOptions, options: copyOptions): ConstantEntity {
        return {
            explanation: instance.explanation,
            constantEntityName: instance.constantEntityName,
            entityShape: instance.entityShape ? instance.entityShape.copyData(resOpt, options) as (string | EntityReferenceDefinition) : undefined,
            constantValues: instance.constantValues
        };
    }
}

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/ConstantEntityPersistence.java
package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmConstantEntityDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.ConstantEntity;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.JMapper;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import java.util.List;

public class ConstantEntityPersistence {

  public static CdmConstantEntityDefinition fromData(final CdmCorpusContext ctx, final JsonNode obj) {
    final String constantEntityName = obj.get("constantEntityName") != null ? obj.get("constantEntityName").asText() : null;
    final CdmConstantEntityDefinition constantEntity = ctx.getCorpus().makeObject(CdmObjectType.ConstantEntityDef, constantEntityName);

    if (obj.get("explanation") != null) {
      constantEntity.setExplanation(obj.get("explanation").asText());
    }

    constantEntity.setConstantValues(JMapper.MAP.convertValue(obj.get("constantValues"), new TypeReference<List<List<String>>>(){}));
    constantEntity.setEntityShape(EntityReferencePersistence.fromData(ctx, obj.get("entityShape")));

    return constantEntity;
  }

  public static ConstantEntity toData(final CdmConstantEntityDefinition instance, final ResolveOptions resOpt, final CopyOptions options) {
    final ConstantEntity constantEntity = new ConstantEntity();

    constantEntity.setExplanation(instance.getExplanation());
    constantEntity.setConstantEntityName(instance.getConstantEntityName());
    constantEntity.setEntityShape(Utils.jsonForm(instance.getEntityShape(), resOpt, options));
    constantEntity.setConstantValues(instance.getConstantValues());

    return constantEntity;
  }
}
